File: Sources/SnapLingo/ProviderClient.swift
import Foundation

struct ProviderOutput: Equatable {
    let text: String
}

struct ProviderPrompt {
    let systemInstruction: String
    let userText: String
}

@MainActor
protocol ProviderClient {
    func translate(_ text: String) async throws -> ProviderOutput
    func polish(_ text: String) async throws -> ProviderOutput
}

enum ProviderValidation {
    static func validate(_ output: ProviderOutput, input: String, mode: TranslationMode) throws -> ProviderOutput {
        let normalized = output.text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty else { throw ProviderError.emptyOutput }

        if mode == .translate, normalized.caseInsensitiveCompare(input.trimmingCharacters(in: .whitespacesAndNewlines)) == .orderedSame {
            throw ProviderError.unusableOutput
        }

        return ProviderOutput(text: normalized)
    }
}

enum ProviderError: LocalizedError {
    case missingAPIKey
    case emptyOutput
    case unusableOutput
    case invalidResponse
    case httpError(Int)
    case unsupportedProvider(String)

    var errorDescription: String? {
        switch self {
        case .missingAPIKey:
            return "Add an API key in Settings to use translation."
        case .emptyOutput:
            return "The provider returned an empty result."
        case .unusableOutput:
            return "The provider returned a result SnapLingo cannot trust."
        case .invalidResponse:
            return "The provider response was malformed."
        case let .httpError(code):
            return "The provider request failed with HTTP \(code)."
        case let .unsupportedProvider(name):
            return "\(name) is not configured correctly."
        }
    }
}

private func loadAPIKey(from credentialStore: CredentialStore, key: CredentialKey) throws -> String {
    do {
        return try credentialStore.loadSecret(for: key)
    } catch CredentialStoreError.notFound {
        throw ProviderError.missingAPIKey
    }
}

private func joinTextParts(_ parts: [[String: Any]]) -> String {
    parts.compactMap { $0["text"] as? String }
        .joined(separator: "\n")
        .trimmingCharacters(in: .whitespacesAndNewlines)
}

@MainActor
struct OpenAICompatibleResponsesProvider: ProviderClient {
    private let preset: ProviderPreset
    private let credentialStore: CredentialStore
    private let session: URLSession

    init(preset: ProviderPreset, credentialStore: CredentialStore, session: URLSession = .shared) {
        self.preset = preset
        self.credentialStore = credentialStore
        self.session = session
    }

    func translate(_ text: String) async throws -> ProviderOutput {
        let output = try await request(prompt: ProviderPrompt(
            systemInstruction: "Translate the user's text into natural, professional English. Return only the final translated text.",
            userText: text
        ))
        return try ProviderValidation.validate(output, input: text, mode: .translate)
    }

    func polish(_ text: String) async throws -> ProviderOutput {
        let output = try await request(prompt: ProviderPrompt(
            systemInstruction: "Rewrite the user's text into natural, professional English that is ready to send. Keep the meaning intact. Return only the final polished text.",
            userText: text
        ))
        return try ProviderValidation.validate(output, input: text, mode: .polish)
    }

    private func request(prompt: ProviderPrompt) async throws -> ProviderOutput {
        let apiKey = try loadAPIKey(from: credentialStore, key: preset.kind.credentialKey)
        let url = URL(string: "\(preset.baseURL)/responses")!

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 20

        let body: [String: Any] = [
            "model": preset.model,
            "instructions": prompt.systemInstruction,
            "input": prompt.userText,
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ProviderError.invalidResponse
        }
        guard 200 ..< 300 ~= httpResponse.statusCode else {
            throw ProviderError.httpError(httpResponse.statusCode)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw ProviderError.invalidResponse
        }

        if let outputText = json["output_text"] as? String, !outputText.isEmpty {
            return ProviderOutput(text: outputText)
        }

        if let output = json["output"] as? [[String: Any]] {
            let chunks = output.flatMap { item -> [String] in
                guard let content = item["content"] as? [[String: Any]] else { return [] }
                return content.compactMap { contentItem in
                    if let text = contentItem["text"] as? String { return text }
                    return nil
                }
            }

            let combined = chunks.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
            if !combined.isEmpty {
                return ProviderOutput(text: combined)
            }
        }

        throw ProviderError.invalidResponse
    }
}

@MainActor
struct OpenAICompatibleChatProvider: ProviderClient {
    private let preset: ProviderPreset
    private let credentialStore: CredentialStore
    private let session: URLSession

    init(preset: ProviderPreset, credentialStore: CredentialStore, session: URLSession = .shared) {
        self.preset = preset
        self.credentialStore = credentialStore
        self.session = session
    }

    func translate(_ text: String) async throws -> ProviderOutput {
        let output = try await request(prompt: ProviderPrompt(
            systemInstruction: "Translate the user's text into natural, professional English. Return only the final translated text.",
            userText: text
        ))
        return try ProviderValidation.validate(output, input: text, mode: .translate)
    }

    func polish(_ text: String) async throws -> ProviderOutput {
        let output = try await request(prompt: ProviderPrompt(
            systemInstruction: "Rewrite the user's text into natural, professional English that is ready to send. Keep the meaning intact. Return only the final polished text.",
            userText: text
        ))
        return try ProviderValidation.validate(output, input: text, mode: .polish)
    }

    private func request(prompt: ProviderPrompt) async throws -> ProviderOutput {
        let apiKey = try loadAPIKey(from: credentialStore, key: preset.kind.credentialKey)
        var request = URLRequest(url: URL(string: "\(preset.baseURL)/chat/completions")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 20

        let body: [String: Any] = [
            "model": preset.model,
            "messages": [
                [
                    "role": "system",
                    "content": prompt.systemInstruction,
                ],
                [
                    "role": "user",
                    "content": prompt.userText,
                ],
            ],
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ProviderError.invalidResponse
        }
        guard 200 ..< 300 ~= httpResponse.statusCode else {
            throw ProviderError.httpError(httpResponse.statusCode)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let choices = json["choices"] as? [[String: Any]],
              let first = choices.first,
              let message = first["message"] as? [String: Any] else {
            throw ProviderError.invalidResponse
        }

        let combined: String
        if let text = message["content"] as? String {
            combined = text.trimmingCharacters(in: .whitespacesAndNewlines)
        } else if let parts = message["content"] as? [[String: Any]] {
            combined = joinTextParts(parts)
        } else {
            throw ProviderError.invalidResponse
        }

        guard !combined.isEmpty else { throw ProviderError.invalidResponse }
        return ProviderOutput(text: combined)
    }
}

@MainActor
struct AnthropicMessagesProvider: ProviderClient {
    private let preset: ProviderPreset
    private let credentialStore: CredentialStore
    private let session: URLSession

    init(preset: ProviderPreset, credentialStore: CredentialStore, session: URLSession = .shared) {
        self.preset = preset
        self.credentialStore = credentialStore
        self.session = session
    }

    func translate(_ text: String) async throws -> ProviderOutput {
        let output = try await request(prompt: ProviderPrompt(
            systemInstruction: "Translate the user's text into natural, professional English. Return only the final translated text.",
            userText: text
        ))
        return try ProviderValidation.validate(output, input: text, mode: .translate)
    }

    func polish(_ text: String) async throws -> ProviderOutput {
        let output = try await request(prompt: ProviderPrompt(
            systemInstruction: "Rewrite the user's text into natural, professional English that is ready to send. Keep the meaning intact. Return only the final polished text.",
            userText: text
        ))
        return try ProviderValidation.validate(output, input: text, mode: .polish)
    }

    private func request(prompt: ProviderPrompt) async throws -> ProviderOutput {
        let apiKey = try loadAPIKey(from: credentialStore, key: preset.kind.credentialKey)
        var request = URLRequest(url: URL(string: "\(preset.baseURL)/messages")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        request.timeoutInterval = 20

        let body: [String: Any] = [
            "model": preset.model,
            "max_tokens": 1024,
            "system": prompt.systemInstruction,
            "messages": [
                [
                    "role": "user",
                    "content": prompt.userText,
                ],
            ],
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ProviderError.invalidResponse
        }
        guard 200 ..< 300 ~= httpResponse.statusCode else {
            throw ProviderError.httpError(httpResponse.statusCode)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let content = json["content"] as? [[String: Any]] else {
            throw ProviderError.invalidResponse
        }

        let combined = content
            .compactMap { item -> String? in
                guard (item["type"] as? String) == "text" else { return nil }
                return item["text"] as? String
            }
            .joined(separator: "\n")
            .trimmingCharacters(in: .whitespacesAndNewlines)

        guard !combined.isEmpty else { throw ProviderError.invalidResponse }
        return ProviderOutput(text: combined)
    }
}

@MainActor
struct GeminiGenerateContentProvider: ProviderClient {
    private let preset: ProviderPreset
    private let credentialStore: CredentialStore
    private let session: URLSession

    init(preset: ProviderPreset, credentialStore: CredentialStore, session: URLSession = .shared) {
        self.preset = preset
        self.credentialStore = credentialStore
        self.session = session
    }

    func translate(_ text: String) async throws -> ProviderOutput {
        let output = try await request(prompt: ProviderPrompt(
            systemInstruction: "Translate the user's text into natural, professional English. Return only the final translated text.",
            userText: text
        ))
        return try ProviderValidation.validate(output, input: text, mode: .translate)
    }

    func polish(_ text: String) async throws -> ProviderOutput {
        let output = try await request(prompt: ProviderPrompt(
            systemInstruction: "Rewrite the user's text into natural, professional English that is ready to send. Keep the meaning intact. Return only the final polished text.",
            userText: text
        ))
        return try ProviderValidation.validate(output, input: text, mode: .polish)
    }

    private func request(prompt: ProviderPrompt) async throws -> ProviderOutput {
        let apiKey = try loadAPIKey(from: credentialStore, key: preset.kind.credentialKey)
        var request = URLRequest(url: URL(string: "\(preset.baseURL)/\(preset.model):generateContent")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-goog-api-key")
        request.timeoutInterval = 20

        let body: [String: Any] = [
            "contents": [
                [
                    "role": "user",
                    "parts": [
                        [
                            "text": "\(prompt.systemInstruction)\n\nUser text:\n\(prompt.userText)",
                        ],
                    ],
                ],
            ],
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ProviderError.invalidResponse
        }
        guard 200 ..< 300 ~= httpResponse.statusCode else {
            throw ProviderError.httpError(httpResponse.statusCode)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let candidates = json["candidates"] as? [[String: Any]],
              let first = candidates.first,
              let content = first["content"] as? [String: Any],
              let parts = content["parts"] as? [[String: Any]] else {
            throw ProviderError.invalidResponse
        }

        let combined = joinTextParts(parts)
        guard !combined.isEmpty else { throw ProviderError.invalidResponse }
        return ProviderOutput(text: combined)
    }
}

File: SnapLingoWindows/Services/ProviderClient.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace SnapLingoWindows.Services;

public interface IProviderClient
{
    Task<ProviderOutput> TranslateAsync(string text, CancellationToken cancellationToken);
    Task<ProviderOutput> PolishAsync(string text, CancellationToken cancellationToken);
}

public sealed class ProviderException : Exception
{
    public ProviderException(string message)
        : base(message)
    {
    }
}

public static class ProviderValidation
{
    public static ProviderOutput Validate(ProviderOutput output, string input, TranslationMode mode, LocalizationService localizer)
    {
        var normalized = output.Text.Trim();
        if (string.IsNullOrWhiteSpace(normalized))
        {
            throw new ProviderException(localizer.Get("error_provider_empty_result"));
        }

        if (mode == TranslationMode.Translate &&
            string.Equals(normalized, input.Trim(), StringComparison.OrdinalIgnoreCase))
        {
            throw new ProviderException(localizer.Get("error_provider_untrusted_result"));
        }

        return new ProviderOutput(normalized);
    }
}

internal abstract class ProviderClientBase : IProviderClient
{
    private static readonly TimeSpan[] RetryDelays =
    [
        TimeSpan.FromMilliseconds(350),
        TimeSpan.FromMilliseconds(900),
    ];

    private readonly ProviderPreset preset;
    private readonly PromptProfile promptProfile;
    private readonly LocalizationService localizer;
    private readonly SecureSecretStore secretStore;
    private readonly HttpClient httpClient;

    protected ProviderClientBase(ProviderPreset preset, PromptProfile promptProfile, LocalizationService localizer, SecureSecretStore secretStore, HttpClient httpClient)
    {
        this.preset = preset;
        this.promptProfile = promptProfile;
        this.localizer = localizer;
        this.secretStore = secretStore;
        this.httpClient = httpClient;
    }

    public abstract Task<ProviderOutput> TranslateAsync(string text, CancellationToken cancellationToken);
    public abstract Task<ProviderOutput> PolishAsync(string text, CancellationToken cancellationToken);

    protected ProviderPreset Preset => preset;
    protected PromptProfile PromptProfile => promptProfile;
    protected LocalizationService Localizer => localizer;

    protected string LoadApiKey()
    {
        var key = secretStore.LoadSecret(preset.Kind);
        if (string.IsNullOrWhiteSpace(key))
        {
            throw new ProviderException(localizer.Get("error_add_api_key"));
        }

        return key.Trim();
    }

    protected async Task<JsonNode> SendAsync(Func<HttpRequestMessage> requestFactory, CancellationToken cancellationToken)
    {
        for (var attempt = 0; ; attempt++)
        {
            using var request = requestFactory();
            using var response = await httpClient.SendAsync(request, cancellationToken);
            var body = await response.Content.ReadAsStringAsync(cancellationToken);

            if (response.IsSuccessStatusCode)
            {
                return JsonNode.Parse(body) ?? throw new ProviderException(localizer.Get("error_provider_malformed"));
            }

            var statusCode = (int)response.StatusCode;
            if (IsTransientStatus(statusCode) && attempt < RetryDelays.Length)
            {
                await Task.Delay(RetryDelays[attempt], cancellationToken);
                continue;
            }

            if (statusCode == 529)
            {
                throw new ProviderException(localizer.Get("error_provider_busy"));
            }

            throw new ProviderException(localizer.Format("error_provider_http", statusCode));
        }
    }

    protected static StringContent JsonContent(object body)
    {
        return new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
    }

    protected static string JoinTextParts(JsonArray? parts)
    {
        if (parts is null)
        {
            return string.Empty;
        }

        return string.Join(
            "\n",
            parts.Select(part => part?["text"]?.GetValue<string>())
                .Where(text => !string.IsNullOrWhiteSpace(text))
        ).Trim();
    }

    private static bool IsTransientStatus(int statusCode)
    {
        return statusCode is 429 or 500 or 502 or 503 or 504 or 529;
    }
}

internal sealed class OpenAIResponsesProvider : ProviderClientBase
{
    public OpenAIResponsesProvider(ProviderPreset preset, PromptProfile promptProfile, LocalizationService localizer, SecureSecretStore secretStore, HttpClient httpClient)
        : base(preset, promptProfile, localizer, secretStore, httpClient)
    {
    }

    public override async Task<ProviderOutput> TranslateAsync(string text, CancellationToken cancellationToken)
    {
        var output = await RequestAsync(PromptProfile.TranslatePrompt, text, cancellationToken);
        return ProviderValidation.Validate(output, text, TranslationMode.Translate, Localizer);
    }

    public override async Task<ProviderOutput> PolishAsync(string text, CancellationToken cancellationToken)
    {
        var output = await RequestAsync(PromptProfile.PolishPrompt, text, cancellationToken);
        return ProviderValidation.Validate(output, text, TranslationMode.Polish, Localizer);
    }

    private async Task<ProviderOutput> RequestAsync(string instructions, string text, CancellationToken cancellationToken)
    {
        HttpRequestMessage CreateRequest()
        {
            var request = new HttpRequestMessage(HttpMethod.Post, $"{Preset.BaseUrl}/responses")
            {
                Content = JsonContent(new
                {
                    model = Preset.Model,
                    instructions,
                    input = text,
                }),
            };
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", LoadApiKey());
            return request;
        }

        var root = await SendAsync(CreateRequest, cancellationToken);

        if (root["output_text"]?.GetValue<string>() is string outputText &&
            !string.IsNullOrWhiteSpace(outputText))
        {
            return new ProviderOutput(outputText.Trim());
        }

        if (root["output"] is JsonArray outputArray)
        {
            var chunks = outputArray
                .SelectMany(item => item?["content"] as JsonArray ?? [])
                .Select(content => content?["text"]?.GetValue<string>())
                .Where(value => !string.IsNullOrWhiteSpace(value));

            var combined = string.Join("\n", chunks).Trim();
            if (!string.IsNullOrWhiteSpace(combined))
            {
                return new ProviderOutput(combined);
            }
        }

        throw new ProviderException(Localizer.Get("error_provider_malformed"));
    }
}

internal sealed class OpenAIChatProvider : ProviderClientBase
{
    public OpenAIChatProvider(ProviderPreset preset, PromptProfile promptProfile, LocalizationService localizer, SecureSecretStore secretStore, HttpClient httpClient)
        : base(preset, promptProfile, localizer, secretStore, httpClient)
    {
    }

    public override async Task<ProviderOutput> TranslateAsync(string text, CancellationToken cancellationToken)
    {
        var output = await RequestAsync(PromptProfile.TranslatePrompt, text, cancellationToken);
        return ProviderValidation.Validate(output, text, TranslationMode.Translate, Localizer);
    }

    public override async Task<ProviderOutput> PolishAsync(string text, CancellationToken cancellationToken)
    {
        var output = await RequestAsync(PromptProfile.PolishPrompt, text, cancellationToken);
        return ProviderValidation.Validate(output, text, TranslationMode.Polish, Localizer);
    }

    private async Task<ProviderOutput> RequestAsync(string instructions, string text, CancellationToken cancellationToken)
    {
        HttpRequestMessage CreateRequest()
        {
            var request = new HttpRequestMessage(HttpMethod.Post, $"{Preset.BaseUrl}/chat/completions")
            {
                Content = JsonContent(new
                {
                    model = Preset.Model,
                    messages = new object[]
                    {
                        new { role = "system", content = instructions },
                        new { role = "user", content = text },
                    },
                }),
            };
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", LoadApiKey());
            return request;
        }

        var root = await SendAsync(CreateRequest, cancellationToken);
        var message = root["choices"]?[0]?["message"];
        if (message is null)
        {
            throw new ProviderException(Localizer.Get("error_provider_malformed"));
        }

        if (message["content"] is JsonValue value &&
            value.TryGetValue<string>(out var stringContent) &&
            !string.IsNullOrWhiteSpace(stringContent))
        {
            return new ProviderOutput(stringContent.Trim());
        }

        if (message["content"] is JsonArray parts)
        {
            var combined = JoinTextParts(parts);
            if (!string.IsNullOrWhiteSpace(combined))
            {
                return new ProviderOutput(combined);
            }
        }

        throw new ProviderException(Localizer.Get("error_provider_malformed"));
    }
}

internal sealed class AnthropicMessagesProvider : ProviderClientBase
{
    public AnthropicMessagesProvider(ProviderPreset preset, PromptProfile promptProfile, LocalizationService localizer, SecureSecretStore secretStore, HttpClient httpClient)
        : base(preset, promptProfile, localizer, secretStore, httpClient)
    {
    }

    public override async Task<ProviderOutput> TranslateAsync(string text, CancellationToken cancellationToken)
    {
        var output = await RequestAsync(PromptProfile.TranslatePrompt, text, cancellationToken);
        return ProviderValidation.Validate(output, text, TranslationMode.Translate, Localizer);
    }

    public override async Task<ProviderOutput> PolishAsync(string text, CancellationToken cancellationToken)
    {
        var output = await RequestAsync(PromptProfile.PolishPrompt, text, cancellationToken);
        return ProviderValidation.Validate(output, text, TranslationMode.Polish, Localizer);
    }

    private async Task<ProviderOutput> RequestAsync(string instructions, string text, CancellationToken cancellationToken)
    {
        HttpRequestMessage CreateRequest()
        {
            var request = new HttpRequestMessage(HttpMethod.Post, $"{Preset.BaseUrl}/messages")
            {
                Content = JsonContent(new
                {
                    model = Preset.Model,
                    max_tokens = 1024,
                    system = instructions,
                    messages = new object[]
                    {
                        new { role = "user", content = text },
                    },
                }),
            };
            request.Headers.Add("x-api-key", LoadApiKey());
            request.Headers.Add("anthropic-version", "2023-06-01");
            return request;
        }

        var root = await SendAsync(CreateRequest, cancellationToken);
        if (root["content"] is not JsonArray contentArray)
        {
            throw new ProviderException(Localizer.Get("error_provider_malformed"));
        }

        var combined = string.Join(
            "\n",
            contentArray
                .Where(item => string.Equals(item?["type"]?.GetValue<string>(), "text", StringComparison.Ordinal))
                .Select(item => item?["text"]?.GetValue<string>())
                .Where(value => !string.IsNullOrWhiteSpace(value))
        ).Trim();

        if (string.IsNullOrWhiteSpace(combined))
        {
            throw new ProviderException(Localizer.Get("error_provider_malformed"));
        }

        return new ProviderOutput(combined);
    }
}

internal sealed class GeminiGenerateContentProvider : ProviderClientBase
{
    public GeminiGenerateContentProvider(ProviderPreset preset, PromptProfile promptProfile, LocalizationService localizer, SecureSecretStore secretStore, HttpClient httpClient)
        : base(preset, promptProfile, localizer, secretStore, httpClient)
    {
    }

    public override async Task<ProviderOutput> TranslateAsync(string text, CancellationToken cancellationToken)
    {
        var output = await RequestAsync(PromptProfile.TranslatePrompt, text, cancellationToken);
        return ProviderValidation.Validate(output, text, TranslationMode.Translate, Localizer);
    }

    public override async Task<ProviderOutput> PolishAsync(string text, CancellationToken cancellationToken)
    {
        var output = await RequestAsync(PromptProfile.PolishPrompt, text, cancellationToken);
        return ProviderValidation.Validate(output, text, TranslationMode.Polish, Localizer);
    }

    private async Task<ProviderOutput> RequestAsync(string instructions, string text, CancellationToken cancellationToken)
    {
        HttpRequestMessage CreateRequest()
        {
            var request = new HttpRequestMessage(HttpMethod.Post, $"{Preset.BaseUrl}/{Preset.Model}:generateContent")
            {
                Content = JsonContent(new
                {
                    contents = new object[]
                    {
                        new
                        {
                            role = "user",
                            parts = new object[]
                            {
                                new { text = $"{instructions}\n\nUser text:\n{text}" },
                            },
                        },
                    },
                }),
            };
            request.Headers.Add("x-goog-api-key", LoadApiKey());
            return request;
        }

        var root = await SendAsync(CreateRequest, cancellationToken);
        var parts = root["candidates"]?[0]?["content"]?["parts"] as JsonArray;
        var combined = JoinTextParts(parts);
        if (!string.IsNullOrWhiteSpace(combined))
        {
            return new ProviderOutput(combined);
        }

        throw new ProviderException(Localizer.Get("error_provider_malformed"));
    }
}
